compiler/rustc_infer/src/infer/combine.rs

```diff
@@ -551,7 +551,8 @@ impl TypeRelation<'tcx> for Generalizer<'_, 'tcx> {
     where
         T: Relate<'tcx>,
     {
-        Ok(ty::Binder::bind(self.relate(a.skip_binder(), b.skip_binder())?))
+        let result = self.relate(a.skip_binder(), b.skip_binder())?;
+        Ok(a.map_bound(|_| result))
```

Comment:
I'm not 100% sure this is correct. It seems to assume that `a` and `b` have equivalent bound variables and hence that the result will be expressed in terms of those variables. This is probably true -- if it's not, we should probably not be folding over binders but instead replacing them with inference variables and the like -- but I feel like I'd like an assertion or something.

Reply:
So, from my other PR, I haven't didn't run into this being a problem. I'm not exactly sure what kind of assertion we could make here.

Reply:
Okay, I am going to revert this here. Once we get to a point that we *can* have an assertion here, we almost certainly want to switch this over.